benchmark/setup.ts

```diff
@@ -507,3 +508,28 @@ function withResolvers<T>() {
 		promise
 	};
 }
+
+export function concurrent<T, U, S extends BenchmarkOptions<T, U>>(suite: S & HasConcurrencyOptions): S {
+	let index = 0;
+	const concurrencyMaximum = suite.concurrencyMaximum ?? 8;
+	let currentlyExecuting: Promise<void>[] = Array(concurrencyMaximum);
+	const restEachTurn = suite.restEachTurn ?? true;
+	return {
+		...suite,
+		async bench(ctx: BenchmarkContext<T>) {
+			await currentlyExecuting[index]; // await the previous execution for this slot
+			currentlyExecuting[index] = suite.bench(ctx) as Promise<void>; // let it execute concurrently and resolve after concurrencyMaximum number of executions
+			index = (index + 1) % concurrencyMaximum; // cycle in a loop
+			if (restEachTurn) { // let asynchronous actions have turns in the event queue, more realistic for most scenarios
+				await rest();
+			}
+		},
+		async teardown(ctx: BenchmarkContext<T>) {
+			// wait for all outstanding executions to finish
+			await Promise.all(currentlyExecuting);
+			// any more tearing down
+			return suite.teardown?.(ctx);
+		}
+	}
+}
+type HasConcurrencyOptions = { concurrencyMaximum?: number, restEachTurn?: boolean };
```

Comment:
Missing `numWorkers`:

```suggestion
type HasConcurrencyOptions = {
	concurrencyMaximum?: number;
	numWorkers?: number;
	restEachTurn?: boolean;
};
```